### DISCUSSION TITLE
A flag or configuration option to soft delete when pruning

### DISCUSSION THREAD

**Discussion by Kyzegs** [laravel/framework#45628] [Ideas]
All records that meet the prunable query are force deleted when running `php artisan model:prune`. It would be great to have an option to disable this behavio either through a flag or config option.

**Comment by Sophist-UK** [comment]
I think an example would help us to understand exactly what the problem and your proposed solution are.

**Reply by Kyzegs** [reply]
It's pretty self-explanatory, I believe. But the `Prunable` traits will force delete any records that meet the `prunable` query. That means it'll still delete the records from the database even if you have the `SoftDelete` trait on your model. It would be great if we had the option to disable this behavior.

Something along the lines of `php aritsan model:prune --soft-delete`.

**Reply by Sophist-UK** [reply]
I am in two minds about this...

On the one hand, the primary idea of pruning is to reduce the size of the database table, so soft-deleting the records rather defeats the point.

But I do see the need to Soft Prune older records in exactly the same way i.e. you might soft-delete records older than a year, and hard prune even older records i.e. hard-delete records older than 2 years.

But this would IMO be separate but similar functionality i.e. you could add a SoftDeletePrune trait or add a softDeletePrune method to the SoftDelete trait.

**Reply by Kyzegs** [reply]
You have a point. I've decided to implement a `SafePrunable` trait within my application. It's a copy of the `Prunable` trait, with the only exception that it does not use `forceDelete` anywhere.

**Reply by Sophist-UK** [reply]
SafePrunable sounds good, but SoftPrunable sounds better because it has a clear connection to SoftDeletes. Also, IMO it should work as Prunable does if the Model doesn't have SoftDelete.

I guess you would also need an `artisan model:softprune` command to match.

**Reply by Sophist-UK** [reply]
P.S. Please consider separating this in to a separate package and publishing it with Packagist.

**Reply by Kyzegs** [reply]
I will make a package for it. I'll probably keep the name as it is. I've also decided that the command will stay `model:prune`. I'll extend Laravel's existing command and adjust the `isPrunable` method to check if the class has the `SafePrunable` trait.

**Reply by Kyzegs** [reply]
You can [click here](https://packagist.org/packages/kyzegs/prunable) to view the package. I still plan on adding a README and tests.
